src/near2far.cpp

```diff
@@ -363,6 +364,84 @@ void dft_near2far::save_farfields(const char *fname, const char *prefix,
     delete[] EH;
 }
 
+double *dft_near2far::flux(direction df, const volume &where, double resolution) {
+
+    /* compute output grid size etc. */
+    int dims[4] = {1,1,1,1};
+    double dx[3] = {0,0,0};
+    direction dirs[3] = {X,Y,Z};
+    int rank = 0, N = 1;
+    LOOP_OVER_DIRECTIONS(where.dim, d) {
+      dims[rank] = int(floor(where.in_direction(d) * resolution));
+      if (dims[rank] <= 1) {
+	dims[rank] = 1;
+	dx[rank] = 0;
+      }
+      else
+	dx[rank] = where.in_direction(d) / (dims[rank] - 1);
+      N *= dims[rank];
+      dirs[rank++] = d;
+    }
+
+    /* 6 x N x Nfreq array of fields in row-major order */
+    std::complex<double> *EH = new std::complex<double>[6*N*Nfreq];
+    std::complex<double> *EH_ = new std::complex<double>[6*N*Nfreq]; // temp array for sum_to_master
+ 
+    /* fields for farfield_lowlevel for a single output point x */
+    std::complex<double> *EH1 = new std::complex<double>[6*Nfreq];
+
+    vec x(where.dim);
+    for (int i0 = 0; i0 < dims[0]; ++i0) {
+      x.set_direction(dirs[0], where.in_direction_min(dirs[0]) + i0*dx[0]);
+      for (int i1 = 0; i1 < dims[1]; ++i1) {
+	x.set_direction(dirs[1],
+			where.in_direction_min(dirs[1]) + i1*dx[1]);
+	for (int i2 = 0; i2 < dims[2]; ++i2) {
+	  x.set_direction(dirs[2],
+			  where.in_direction_min(dirs[2]) + i2*dx[2]);
+	  farfield_lowlevel(EH1, x);
+	  int idx = (i0 * dims[1] + i1) * dims[2] + i2;
+	  for (int i = 0; i < Nfreq; ++i)
+	    for (int k = 0; k < 6; ++k)
+	      EH_[(k * N + idx) + (6 * N * i)] = EH1[i * 6 + k];
+	}
+      }
+    }
+
+    delete[] EH1;
+    sum_to_master(EH_, EH, 6*N*Nfreq);
+    delete[] EH_;
+
+    if (coordinate_mismatch(where.dim, df) || where.dim == Dcyl)
+      abort("cannot get flux for near2far: co-ordinate mismatch");
+
+    double *F = new double[Nfreq];
+    double *F_ = new double[Nfreq]; // temp array for sum to master
+    std::complex<double> *ff_EH[6];
+    std::complex<double> *cE[2], *cH[2];
+    for (int i = 0; i < Nfreq; ++i) {
+      for (int k = 0; k < 6; ++k)
+	ff_EH[k] = EH + (k * N) + (6 * N * i);
+      switch (df) {
+      case X: cE[0] = ff_EH[1], cE[1] = ff_EH[2], cH[0] = ff_EH[5], cH[1] = ff_EH[4]; break;
+      case Y: cE[0] = ff_EH[2], cE[1] = ff_EH[0], cH[0] = ff_EH[3], cH[1] = ff_EH[5]; break;
+      case Z: cE[0] = ff_EH[0], cE[1] = ff_EH[1], cH[0] = ff_EH[4], cH[1] = ff_EH[3]; break;
+      case NO_DIRECTION: abort("cannot get flux in NO_DIRECTION");
+      }
+      F_[i] = 0;
+      for (int j = 0; j < 2; ++j) {
+        double flux_sum = 0;
+	for (int p = 0; p < N; ++p)
+	  flux_sum += real(cE[j][p]*conj(cH[j][p]));
+	F_[i] += flux_sum * (1 - 2*j);
+      }
+      master_printf("near2far-flux:, %0.2g, %0.5g\n",freq_min+i*dfreq,F_[i]);
+    }
+    sum_to_all(F_, F, Nfreq);
```

Comment:
This `sum_to_all` seems redundant with the `sum_to_master` above?  Also, isn't there a memory leak because you're not freeing `EH`?